pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go

```diff
@@ -283,6 +414,20 @@ func addColumnsForSecondaryIndex(
 			e.Kind != scpb.IndexColumn_KEY {
 			return
 		}
+		// Check if the column name was duplicated from the STORING clause, in which
+		// case this isn't allowed.
+		// Note: The column IDs for the key suffix columns are not resolved, so we couldn't
+		// do this earlier.
+		scpb.ForEachColumnName(relationElements, func(current scpb.Status, target scpb.TargetStatus, cn *scpb.ColumnName) {
+			if cn.ColumnID == e.ColumnID &&
+				cn.TableID == e.TableID {
+				if _, found := columnRefs[cn.Name]; found {
+					panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
+						"index %q already contains column %q", n.Name, cn.Name))
+				}
+				columnRefs[cn.Name] = struct{}{}
+			}
+		})
```

Comment:
1. Why do we have to check "whether there is duplicate STORING columns" here, but not earlier? I didn't understand the comment "Note: The column IDs for the key suffix columns are not resolved, so we couldn't do this earlier."

2. You are computing the name of column `e`, right? If so, how is this related to STORING columns in the to-be-added index as `e` is a key suffix column in the to-be-added index?

3. I think you can just call function `mustRetrieveColumnNameElem` to get the name of a column by specifying table ID and column ID.

Reply:
1) The key suffix columns are derived from the primary key, pretty much any column that isn't contained by the key columns in the index. To make this clearer I'll change this comment to:
// Note: The column IDs for the key suffix columns are derived by finding columns in the primary index, which
// are not covered by the key columns within the index. We checked against the key columns vs the storing ones,
// earlier so this covers any extra columns.

2) If the column is already in the key either implicitly via the suffix or directly by the references key columns, then it makes no sense for it to be stored again. So, this check is pretty much saying the set of storing columns and key suffix columns don't overlap.

3) TIL, let me replace of these